File: springbootDynamoDB-sdk-v2/src/main/java/com/dynamodb/dto/Artist.java
package com.dynamodb.dto;

import com.fasterxml.jackson.annotation.JsonInclude;
import software.amazon.awssdk.enhanced.dynamodb.mapper.annotations.*;
import software.amazon.awssdk.services.dynamodb.model.AttributeValue;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

@DynamoDbBean
@JsonInclude(JsonInclude.Include.NON_EMPTY)
public class Artist {
	
	private String pk;
	private String sk;
	private String gsi1pk;
	private String gsi1sk;
	private String name;
	private String nationality;
	private List<Song> songs = new ArrayList<>();
	
	public Artist() {
	}

	public Artist(String name, String nationality) {
		this.name = name;
		this.nationality = nationality;
	}

	@DynamoDbPartitionKey
	public String getPk() {
		return pk;
	}

	public void setPk(String pk) {
		this.pk = pk;
	}
	
	@DynamoDbSortKey
	public String getSk() {
		return sk;
	}

	public void setSk(String sk) {
		this.sk = sk;
	}

	@DynamoDbSecondaryPartitionKey(indexNames = {"gsi1"})
	public String getGsi1pk() {
		return gsi1pk;
	}

	public void setGsi1pk(String gsi1pk) {
		this.gsi1pk = gsi1pk;
	}
	
	@DynamoDbSecondarySortKey(indexNames = {"gsi1"})
	public String getGsi1sk() {
		return gsi1sk;
	}

	public void setGsi1sk(String gsi1sk) {
		this.gsi1sk = gsi1sk;
	}

	public String getName() {
		return name;
	}
	
	public void setName(String name) {
		this.name = name;
	}

	public String getNationality() {
		return nationality;
	}
	
	public void setNationality(String nationality) {
		this.nationality = nationality;
	}
	
	@DynamoDbIgnore
	public List<Song> getSongs() {
		return songs;
	}
	
	public void setSongs(List<Song> songs) {
		this.songs = songs;
	}

	public static Artist attributeMapToArtist(Map<String, AttributeValue> attributeMap) {
		String name = attributeMap.get("ArtistName").s();
		String nationality = attributeMap.get("Nationality").s();
		return new Artist(name, nationality);
	}

	@Override
	public String toString() {
		return "Artist [name=" + name + ", nationality=" + nationality + ", songs=" + songs + "]";
	}

}

File: springbootDynamoDB-kotlin-sdk-v2/src/main/kotlin/me/brunosantana/dto/Artist.kt
package me.brunosantana.dto

import com.fasterxml.jackson.annotation.JsonInclude
import me.brunosantana.converter.CustomBooleanAttributeConverter
import software.amazon.awssdk.enhanced.dynamodb.mapper.annotations.DynamoDbAttribute
import software.amazon.awssdk.enhanced.dynamodb.mapper.annotations.DynamoDbBean
import software.amazon.awssdk.enhanced.dynamodb.mapper.annotations.DynamoDbConvertedBy
import software.amazon.awssdk.enhanced.dynamodb.mapper.annotations.DynamoDbIgnore
import software.amazon.awssdk.services.dynamodb.model.AttributeValue

private const val TYPE = "ARTIST"

@DynamoDbBean
@JsonInclude(JsonInclude.Include.NON_EMPTY)
data class Artist(
    @get:DynamoDbAttribute("ArtistName")
    var name: String,
    @get:DynamoDbAttribute("Nationality")
    var nationality: String,
    @get:DynamoDbAttribute("IsAwardWinner")
    var isAwardWinner: Boolean,
    @get:DynamoDbAttribute("IsWellKnown")
    @get:DynamoDbConvertedBy(CustomBooleanAttributeConverter::class)
    var isWellKnown: Boolean,
    @get:DynamoDbAttribute("IsAnInternationalSinger")
    var internationalSinger: Boolean? = null,
    @get:DynamoDbAttribute("IsAlsoActor")
    var actor: Int = 0,
    @get:DynamoDbIgnore
    val songs: MutableList<Song> = mutableListOf()
): DynamoBaseModel(
    pkType = "artist",
    pkId = name.lowercase().replace(" ", "_"),
    sk = "artist#${name.lowercase().replace(" ", "_")}",
    gsi1pk = "type#artist",
    gsi1sk = "type#artist",
) {
    constructor(
        name: String,
        nationality: String,
        isAwardWinner: Boolean,
        isWellKnown: Boolean,
        internationalSinger: Boolean?,
        actor: Int
    ) :
            this(
                name = name,
                nationality = nationality,
                isAwardWinner = isAwardWinner,
                isWellKnown = isWellKnown,
                internationalSinger = internationalSinger,
                actor = actor,
                songs = mutableListOf()
            )

    @Deprecated(message = "Intended to be used only by AWS SDK")
    constructor() :
            this(
                name = "",
                nationality = "",
                isAwardWinner = false,
                isWellKnown = true,
                internationalSinger = null,
                actor = 0,
                songs = mutableListOf()
            )

    @DynamoDbAttribute("Type")
    fun getType(): String {
        return TYPE
    }

    fun setType(type: String) {
        // Do nothing, this setter is just to make the AWS SDK 2.x happy
    }

    fun addAllSongs(songs: MutableList<Song>){
        this.songs.addAll(songs)
    }

    companion object {
        fun attributeMapToArtist(attributeMap: Map<String, AttributeValue>): Artist {
            val name = attributeMap["ArtistName"]!!.s()
            val nationality = attributeMap["Nationality"]!!.s()
            val isAwardWinner = attributeMap["IsAwardWinner"]?.bool() ?: false
            val isWellKnownNumber = attributeMap["IsWellKnown"]?.n()
            val isWellKnown = if(isWellKnownNumber != null) isWellKnownNumber.toInt() != 0 else false
            val internationalSinger = attributeMap["IsAnInternationalSinger"]?.bool()
            val actor = attributeMap["IsAlsoActor"]?.n()?.toInt() ?: 0
            val versionTimestamp = attributeMap["VersionTimestamp"]?.s()
            val artist = Artist(
                name = name,
                nationality = nationality,
                isAwardWinner = isAwardWinner,
                isWellKnown = isWellKnown,
                internationalSinger = internationalSinger,
                actor = actor)
            artist.versionTimestamp = versionTimestamp
            return artist
        }
    }
}
